tests/bsim/bluetooth/host/security/ltk_request_hook/test_ltk_request_hook.c

```diff
@@ -0,0 +1,343 @@
+#include <errno.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+#include <sys/types.h>
+#include <testlib/adv.h>
+#include <testlib/att_read.h>
+#include <testlib/att_write.h>
+#include <testlib/conn.h>
+#include <testlib/enable_quiet.h>
+#include <testlib/log_utils.h>
+#include <testlib/scan.h>
+#include <testlib/security.h>
+#include <zephyr/bluetooth/addr.h>
+#include <zephyr/bluetooth/att.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/bluetooth/gatt.h>
+#include <zephyr/bluetooth/hci_types.h>
+#include <zephyr/bluetooth/hci.h>
+#include <zephyr/bluetooth/uuid.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log_core.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/net_buf.h>
+#include <zephyr/settings/settings.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util_macro.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/toolchain.h>
+
+LOG_MODULE_REGISTER(main, LOG_LEVEL_DBG);
+
+#define UUID_1                                                                                     \
+	BT_UUID_DECLARE_128(0xdb, 0x1f, 0xe2, 0x52, 0xf3, 0xc6, 0x43, 0x66, 0xb3, 0x92, 0x5d,      \
+			    0xc6, 0xe7, 0xc9, 0x59, 0x9d)
+
+#define UUID_2                                                                                     \
+	BT_UUID_DECLARE_128(0x3f, 0xa4, 0x7f, 0x44, 0x2e, 0x2a, 0x43, 0x05, 0xab, 0x38, 0x07,      \
+			    0x8d, 0x16, 0xbf, 0x99, 0xf1)
+
+#define UUID_3                                                                                     \
+	BT_UUID_DECLARE_128(0x06, 0x30, 0xbb, 0xae, 0xff, 0x9a, 0x4e, 0x83, 0xa6, 0x5c, 0xf0,      \
+			    0x4e, 0xdf, 0xb8, 0x79, 0x1d)
+
+static ssize_t read_mtu_validation_chrc(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+					void *buf, uint16_t buf_len, uint16_t offset)
+{
+	ARG_UNUSED(conn);
+	ARG_UNUSED(attr);
+	ARG_UNUSED(buf);
+	ARG_UNUSED(buf_len);
+	ARG_UNUSED(offset);
+
+	return 0;
+}
+
+/* Sample GATT service with two characteristics:
+ * - One that requires encryption (no authentication)
+ * - One that requires authentication
+ *
+ * Both characteristics share the same read handler, which successfully returns
+ * an empty payload.
+ */
+static struct bt_gatt_attr attrs[] = {
+	BT_GATT_PRIMARY_SERVICE(UUID_1),
+	BT_GATT_CHARACTERISTIC(UUID_2, BT_GATT_CHRC_READ, BT_GATT_PERM_READ_ENCRYPT,
+			       read_mtu_validation_chrc, NULL, NULL),
+	BT_GATT_CHARACTERISTIC(UUID_3, BT_GATT_CHRC_READ, BT_GATT_PERM_READ_AUTHEN,
+			       read_mtu_validation_chrc, NULL, NULL),
+};
+
+static struct bt_gatt_service sample_svc_requiring_encryption = {
+	.attrs = attrs,
+	.attr_count = ARRAY_SIZE(attrs),
+};
+
+/* Minimal implementation of an LTK request hook.
+ *
+ * The hook allows the application to inspect the HCI LTK Request event and
+ * choose whether the stack should handle it via SMP, or whether the
+ * application will provide the LTK directly at the HCI level.
```

Comment:
This doc is outdated. I'll fix it. It's a callback now, not a hook, and it does not inspect HCI nor decide whether SMP will handle it. SMP has priority now. It is just queried for a LTK by the Host if SMP did not have one, and it has the opportunity to asynchronously reply.